File: Data Structures & Algorithms/combination-target-sum/submission-4.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        temp = []

        def generate_combination(rem: int, ind: int):
            if rem == 0:
                ans.append(temp[:])
                return
            
            if rem < 0 or ind == len(nums):
                return
            
            temp.append(nums[ind])
            generate_combination(rem - nums[ind], ind)
            temp.pop()

            generate_combination(rem, ind + 1)
        
        generate_combination(target, 0)
        return ans

File: Data Structures & Algorithms/subsets-ii/submission-4.ts
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsetsWithDup(nums: number[]): number[][] {
        nums.sort((a, b) => a - b);
        const sol: number[][] = [];

        const allSubsets = (ind: number, temp: number[]): void => {
            sol.push([...temp]);

            for(let i = ind; i < nums.length; ++i) {
                if(ind < i && nums[i] === nums[i - 1]) continue;

                temp.push(nums[i]);
                allSubsets(i + 1, temp);
                temp.pop();
            }
        }

        allSubsets(0, []);
        return sol;
    }
}
